### PR TITLE
[AMD] Optimize gpt-oss-120B performance

### PR DESCRIPTION
**Pull request by kkHuang-amd**
## Motivation

Integrate Atom's `pa_decode_gluon` paged-attention decode kernel into
SGLang's AITER backend to accelerate gpt-oss-120b decode on ROCm
(MI300 / MI355).  The kernel requires a SHUFFLE 5D vectorized KV
cache layout that aiter's `mha_batch_prefill_func` also consumes
natively, so we add the layout as a new opt-in pool format
(`SGLANG_AITER_KV_CACHE_LAYOUT=vectorized_5d`) and route the AITER backend
through it on both prefill and decode.

End result on gpt-oss-120b @ MI300, TP=1, page=64, c=128:

- **+14 % TPOT** and **+15 % output throughput** vs the NHD baseline
  on 8K/1K decode; pa_decode_gluon itself is ~46 % faster than the
  legacy `unified_attention` decode kernel.
- GSM8K (lm_eval, 3-shot, chat template) stays within seed noise of
  baseline for both bf16 and fp8 KV cache.

The layout is fully opt-in — when `SGLANG_AITER_KV_CACHE_LAYOUT` is unset
(default `"nhd"`) every code path is bit-for-bit identical to current
`main`.

### New env var
- `SGLANG_AITER_KV_CACHE_LAYOUT` (`environ.py`) selects the physical KV pool
  layout. `"nhd"` (default) keeps today's behaviour; `"vectorized_5d"`
  enables the SHUFFLE 5D layout.

### Memory pool (`mem_cache/`)
- `MHATokenToKVPool` allocates the 5D layout when the flag is set:
  - K: `(num_blocks, H_kv, head_dim/X, page_size, X)`
  - V: `(num_blocks, H_kv, page_size/X, head_dim, X)`,
    where `X = 16 / store_dtype.itemsize` (8 for bf16, 16 for fp8).
- `SWAKVPool` propagates the layout to both full-attn and SWA sub-pools
  so SWA decode can also go through `pa_decode_gluon`.
- `set_kv_buffer` already does per-tensor fp8 quant on the host before
  the writer kernel, so the kernel just byte-copies into the cache —
  no scale plumbing needed on the writer side.

### Writer kernel (`layers/attention/utils.py`)
- `reshape_and_cache_shuffle_5d` / `launch_reshape_and_cache_shuffle_5d`:
  Triton scatter from per-token `(num_tokens, num_heads, head_size)`
  K/V into the 5D SHUFFLE layout. Used by `MHATokenToKVPool.set_kv_buffer`
  when the fused QKV+RoPE+set_kv writer is unavailable.

### Reader kernel (`layers/attention/utils.py`)
- `gather_shuffle_5d_to_linear` / `launch_gather_shuffle_5d_to_linear`:
  bit-exact triton inverse of the writer. Materialises per-token
  `(T, num_heads, head_size)` K/V from a list of pool slot ids. Used
  by the chunked-prefill gather-and-linearize fallback (see below).

### AITER backend wiring (`layers/attention/aiter_backend.py` + new
`layers/attention/aiter_utils.py`)

Two 5D-specific entry points live in `aiter_utils.py` and are called
from `forward_extend` / `forward_decode` only when the active pool is
5D, keeping the backend file focused on dispatch and the legacy NHD
path:

- `forward_extend_vectorized_5d`
  - **Fresh-prompt shortcut**: when every request in the batch has
    `extend_prefix_lens == 0` (first prefill chunk, no prefix reuse)
    skip pool reads entirely and feed the raw `(k, v)` to
    `mha_batch_prefill_func` in 3D LINEAR mode (page_size=1) on bf16.
  - **Gather-and-linearize**: otherwise gather the full per-request
    K/V from the SHUFFLE 5D pool via `launch_gather_shuffle_5d_to_linear`
    into a contiguous `(T, H, D)` buffer in `store_dtype`, then run
    the same LINEAR prefill. fp8-store layers forward the raw fp8
    tensors with per-tensor descales — aiter's LINEAR-mode prefill
    supports fp8 K/V/Q natively, no host-side dequant needed.
  - Without this path aiter's paged `mha_batch_prefill_func` aborts
    with `"no matching kernel found"` for our
    `(page_size=64, bf16/fp8, SHUFFLE 5D)` configuration whenever the
    scheduler produces a non-first chunked-prefill batch.

- `forward_decode_vectorized_5d`
  - Routes BOTH full-attn and SWA layers through `pa_decode_gluon`
    (SWA sub-pool is also allocated 5D). Full layers pass
    `kv_indices` + `sliding_window=0`; SWA layers pass
    `swa_page_table` + `sliding_window=layer.sliding_window_size`.
    fp8 KV cache forwards `layer.k_scale` / `layer.v_scale` as
    `key_scale` / `value_scale` so the kernel dequantises correctly.

### Fused RoPE + KV write (`layers/rotary_embedding/base.py`, `models/utils.py`)
- The HIP fused `fused_qk_rope_reshape_and_cache` kernel already
  supports the 5D SHUFFLE layout (`flash_layout=False`). Re-enable
  `enable_fused_set_kv_buffer` for the 5D pool and have the arg
  builder pass the raw 5D buffer when `k_buffer.ndim == 5`. This
  collapses RoPE + KV write into a single launch per layer per step.

### UX
- `server_args._handle_page_size` defaults `page_size=64` when
  `SGLANG_AITER_KV_CACHE_LAYOUT=vectorized_5d`, since that is the page size
  the 5D kernels are tuned for.

## Accuracy Tests

All numbers below are GSM8K (3-shot, chat template) on gpt-oss-120b @
MI300, **TP=1**, run with the following commands:

```bash
# Server (vary --kv-cache-dtype between auto / fp8_e4m3; drop the
# SGLANG_AITER_KV_CACHE_LAYOUT env for the NHD baseline rows).
SGLANG_AITER_KV_CACHE_LAYOUT=vectorized_5d \
SGLANG_USE_AITER_MOE_GU_ITLV=False \
python3 -m sglang.launch_server \
  --model-path /path/to/gpt-oss-120b/ \
  --tp 1 --trust-remote-code \
  --mem-fraction-static 0.85 \
  --prefill-attention-backend aiter --decode-attention-backend aiter \
  --chunked-prefill-size 131072 --max-running-requests 256 \
  --page-size 64 --disable-radix-cache --port 8000 \
  --kv-cache-dtype fp8_e4m3   # or `auto` for bf16

# Client
lm_eval --model local-chat-completions --apply_chat_template \
  --model_args model=/path/to/gpt-oss-120b/,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=65,max_retries=3,max_gen_toks=2048,tokenized_requests=False \
  --tasks gsm8k --num_fewshot 3
```

| Layout              | KV dtype  | flexible-extract | strict-match |
|---------------------|-----------|-----------------:|-------------:|
| NHD (baseline)      | bf16      | 0.8855           | 0.3647       |
| NHD                 | fp8_e4m3  | 0.8893           | 0.3381       |
| **SHUFFLE 5D (this PR)** | bf16 | **0.8870**       | **0.3389**   |
| **SHUFFLE 5D (this PR)** | fp8_e4m3 | **0.8848** | **0.3374**   |

Specific path coverage:
- Chunked-prefill gather-and-linearize path (forced by the repro in
  the bugfix commit): bf16 → 0.8863 / 0.3442, fp8 → 0.8848 / 0.3374.
- fp8 with the fused write path disabled (forces the standalone 5D
  writer): 0.8848 / 0.3480 — confirms `set_kv_buffer`'s host-side
  per-tensor quant feeds the writer correctly.

> **TP > 1 caveat:** GSM8K at TP=2 / TP=4 fails on this branch *and*
> on pre-Plan-A `main` (the NHD baseline also drops to ~0.32 / 0.16 at
> TP=2). Suspected upstream MoE FlyDSL kernel dispatch bug at
> per-rank shapes. TP=1 and TP=8 are unaffected. Tracked in
> `KNOWN_ISSUES.md` (A1, local doc) — not introduced by this PR.

## Speed Tests and Profiling

InferenceX-style sweep on 8K/1K decode (256 prompts, MI300, TP=1,
`--page-size 64 --kv-cache-dtype auto --disable-radix-cache`):

| Concurrency | Layout | Median TPOT (ms) | Output throughput (tok/s) |
|--:|---|--:|--:|
|  4 | NHD baseline       | ~4.75 |   780 |
|  4 | SHUFFLE 5D (this)  | **~4.17 (-12 %)** | **~920 (+18 %)** |
|  8 | NHD baseline       | ~6.06 | 1220  |
|  8 | SHUFFLE 5D (this)  | **~4.88 (-19 %)** | **~1530 (+26 %)** |
| 128 | NHD baseline      | ~7.05 | 2007  |
| 128 | SHUFFLE 5D (this) | **~6.06 (-14 %)** | **~2307 (+15 %)** |

For full-context fp8 + 8K c=64 we re-verified single-run noise: the
SHUFFLE 5D + `pa_decode_gluon` path matches ATOM throughput within
±6 % across the sweep and beats it by 18-26 % at low concurrency.

`pa_decode_gluon` itself is ~46 % faster than `unified_attention` on
the same `(bs, num_kv_heads, head_dim, ctx_len)` for our gpt-oss
shapes.

## Known issues / how to use

```bash
SGLANG_AITER_KV_CACHE_LAYOUT=vectorized_5d \
SGLANG_USE_AITER_MOE_GU_ITLV=False \
python3 -m sglang.launch_server \
  --model-path /path/to/gpt-oss-120b/ \
  --tp 1 --trust-remote-code \
  --mem-fraction-static 0.85 \
  --prefill-attention-backend aiter --decode-attention-backend aiter \
  --page-size 64 --disable-radix-cache \
  --max-running-requests 256 \
  --kv-cache-dtype fp8_e4m3
```

- `--max-running-requests 256` (or omit; auto-tuned default is
  >= 256) is **required** for full accuracy on gpt-oss-120b. With
  `--max-running-requests < 256` the AITER MoE dispatch in
  `aiter/fused_moe.py` selects a bf16-activation path (governed by
  `GPTOSS_SWIGLU_MXFP4_BF16_BOUND`, default 256) that produces
  measurably wrong logits on gpt-oss MXFP4 — independently of this
  PR. Workaround for smaller batches:
  `export GPTOSS_SWIGLU_MXFP4_BF16_BOUND=128` (don't set to 0 — that
  crashes the server at init since warmup hits tiny `M`).

## Checklist

* [x] Format your code according to the Format code with pre-commit.
* [ ] Add unit tests according to the Run and add unit tests.
* [x] Update documentation according to Write documentations.
* [x] Provide accuracy and speed benchmark results according to Test
      the accuracy and Benchmark the speed.
* [x] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #27113289949](https://github.com/sgl-project/sglang/actions/runs/27113289949)
Latest PR Test (Extra): :x: [Run #27113289843](https://github.com/sgl-project/sglang/actions/runs/27113289843)

### CODE REVIEW

**Comment by HaiShaw** [comment]
@amd-bot ci-status

**Comment by kkHuang-amd** [comment]
@amd-bot ci-status

**Comment by kkHuang-amd** [comment]
@amd-bot ci-status

File: python/sglang/srt/models/utils.py [hunk 1]
```diff
@@ -275,7 +275,15 @@ def load_weights(
 
 
 def enable_fused_set_kv_buffer(forward_batch: ForwardBatch):
-    """Enable fused set_kv_buffer only on CUDA with bfloat16 KV cache."""
+    """Enable fused set_kv_buffer only on CUDA with bfloat16 KV cache.
```
Comment (HaiShaw) [inline python/sglang/srt/models/utils.py]: should we change the first line comment to include HIP?

File: python/sglang/srt/models/utils.py [hunk 2]
```diff
@@ -313,16 +321,26 @@ def create_fused_set_kv_buffer_arg(
             if layer.sliding_window_size > 0
             else None
         )
+        # SHUFFLE 5D pools (k_buffer.ndim == 5) consumed natively by
```
Comment (HaiShaw) [inline python/sglang/srt/models/utils.py:324]: should we do within  _is_hip or _use_aiter here?

File: python/sglang/srt/environ.py [hunk 3]
```diff
@@ -407,6 +407,26 @@ class Envs:
...
+    # Fuse the `residual_add + RMSNorm + zero-pad` triplet that appears
+    # before the MoE block for models whose MoE input hidden_size must be
+    # padded up to a stride (e.g. GPT-OSS MXFP4 needs pad to multiple of
+    # 256). When False (default) the pad runs as a separate
+    # torch.nn.functional.pad call inside the MoE method. When True, the
+    # aiter Triton kernel `fused_add_rmsnorm_pad` produces a padded
+    # post-attention layernorm output in one launch and the MoE method
+    # skips the explicit pad. Currently only takes effect on the
+    # post_attention_layernorm path with aiter backend and TP=1.
+    SGLANG_FUSE_RMSNORM_PAD = EnvBool(False)
```
Comment (HaiShaw) [inline python/sglang/srt/environ.py]: should this be a HIP sepcific flag?

**Review by HaiShaw — COMMENTED** [review]
Can we apply `_use_aiter` more consistently cross code changes - producer, consumers, utilities?

File: python/sglang/srt/models/utils.py (hunk 2 above)
Comment (kkHuang-amd) [reply python/sglang/srt/models/utils.py:324]: we have a condition if not _is_hip at line 309, so the else branch will all be _is_hip can be in

**Comment by HaiShaw** [comment]
@amd-bot ci-status

**Review by HaiShaw — COMMENTED** [review]
LGTM

File: python/sglang/srt/models/gpt_oss.py [hunk 4]
```diff
@@ -250,18 +282,44 @@ def forward_normal(
...
+        # `hidden_states` may arrive pre-padded along the last dim when the
+        # preceding RMSNorm fused the MoE input pad (gated by
+        # SGLANG_AITER_FUSE_RMSNORM_PAD). Router/topk are computed on the
+        # unpadded slice so the small bf16 router GEMM dimensions stay
+        # untouched, while the experts call gets to keep the padded view
+        # and skip the duplicate pad inside the MXFP4 method. The output
+        # is then trimmed back to the unpadded width so postprocess_layer
+        # can pair it with the (M, hidden_dim_unpadded) residual.
+        num_tokens = hidden_states.shape[0]
+        hidden_dim_unpadded = self.experts.hidden_size
```
Comment (mmangkad) [inline python/sglang/srt/models/gpt_oss.py:294]: This contradicts the comment: `hidden_dim_unpadded` should be the model hidden size (`config.hidden_size`). This regresses GPT-OSS with FlashInfer MXFP4 on SM10X because `self.experts.hidden_size` is padded to 3072, while `config.hidden_size` is 2880.

**Comment by mmangkad** [comment]
@kkHuang-amd @HaiShaw this broke GPT-OSS serving on NVIDIA SM10X with FlashInfer MXFP4. Instead of reverting, this should be a quick fix; I have it here, please check: https://github.com/sgl-project/sglang/pull/27528

**Comment by fzyzcjy** [comment]
https://github.com/sgl-project/sglang/pull/25647#issuecomment-4646429164

**Comment by yuan-luo** [comment]
> @kkHuang-amd @HaiShaw this broke GPT-OSS serving on NVIDIA SM10X with FlashInfer MXFP4. Instead of reverting, this should be a quick fix; I have it here, please check: #27528

I also encountered this error in CI of https://github.com/sgl-project/sglang/pull/27449

**Comment by geraldstanje1** [comment]
hi any plans to make gpt-oss-20b faster on nvidia rtx 6000 pro?
